ArdusatSDK.cpp

```diff
@@ -453,266 +217,1101 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
+ */
+const char * valuesToJSON(const char *sensorName, unsigned char unit, int numValues, ...) {
+  int i = 0;
+  size_t nameLength = strlen(sensorName) + 1; // For null terminator
+  va_list args;
 
-  dtostrf(input.y, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-  _getOutBuf()[_output_buf_len++] = CSV_SEPARATOR;
+  char concatBuf[5];
+  strcpy_P(concatBuf, concat_string);
 
-  dtostrf(input.z, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
+  _resetOutBuf();
+  va_start(args, numValues);
+  for (i = 0; i < numValues; ++i) {
+    char * label = va_arg(args, char *);
+    float value = va_arg(args, double);
+    char nameBuf[nameLength + strlen(label) + 1];
 
-  _add_checksum_to_csv_buffer(sensorName, 3, input.x, input.y, input.z);
-  _getOutBuf()[_output_buf_len++] = '\n';
+    sprintf(nameBuf, concatBuf, sensorName, label); // concatBuf = "%s%s";
+    _writeJSONValue(&_getOutBuf()[_output_buf_len], nameBuf, unit_to_str(unit), value);
+  }
+  va_end(args);
 
   return _getOutBuf();
 }
 
-_DEFAULT_SENSOR_NAME_TO_CSV_FXN(temperature)
-const char * temperatureToCSV(const char *sensorName, temperature_t & input) {
-  _headerToCSV(&(input.header), sensorName);// warning : resets the internal buffer
-
-  dtostrf(input.t, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-
-  _add_checksum_to_csv_buffer(sensorName, 1, input.t);
-  _getOutBuf()[_output_buf_len++] = '\n';
-
+/**
+ * Creates a JSON string with the appropriate values
+ *
+ * @param sensorName the label to be given to some data
+ * @param unit the unit of measurement used
+ * @param value the sensor value
+ *
+ * @return the output buffer
+ */
+const char * valueToJSON(const char *sensorName, unsigned char unit, float value) {
+  _resetOutBuf();
+  _writeJSONValue(_getOutBuf(), sensorName, unit_to_str(unit), value);
   return _getOutBuf();
 }
 
-_DEFAULT_SENSOR_NAME_TO_CSV_FXN(gyro)
-const char * gyroToCSV(const char *sensorName, gyro_t & input) {
-  _headerToCSV(&(input.header), sensorName);// warning : resets the internal buffer
-
-  dtostrf(input.x, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-  _getOutBuf()[_output_buf_len++] = CSV_SEPARATOR;
 
-  dtostrf(input.y, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-  _getOutBuf()[_output_buf_len++] = CSV_SEPARATOR;
+/**************************************************************************//**
+ * @brief   Initializes the sensor with any set configurations
+ * @ingroup sensor
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ *****************************************************************************/
+boolean Sensor::begin(void) {
+  catchSpaceboard();
+  this->initialized = this->initialize();
+
+  if (!this->initialized) {
+    char err_msg[30];
+    char sensor[50];
+
+    strcpy_P(err_msg, begin_error_msg);
```

Comment:
Basically. Except _writeUnavailableSensorError has 3 `char *` variables instead of 2 to `sprintf` into a buffer